redskyapi/experiments/v1alpha1/parameter.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/redskyops/redskyops-controller/redskyapi/experiments/v1alpha1/numstr"
+)
+
+// LowerBound attempts to return the lower bound for this parameter.
+func (p *Parameter) LowerBound() (*numstr.NumberOrString, error) {
+	if p.Type == ParameterTypeCategorical {
+		if len(p.Values) == 0 {
+			return nil, fmt.Errorf("unable to determine categorical minimum bound")
+		}
+		return &numstr.NumberOrString{StrVal: p.Values[0], IsString: true}, nil
+	}
+
+	if p.Bounds == nil {
+		return nil, fmt.Errorf("unable to determine numeric minimum bound")
+	}
+
+	return &numstr.NumberOrString{NumVal: p.Bounds.Min}, nil
+}
+
+// UpperBound attempts to return the upper bound for this parameter.
+func (p *Parameter) UpperBound() (*numstr.NumberOrString, error) {
+	if p.Type == ParameterTypeCategorical {
+		if len(p.Values) == 0 {
+			return nil, fmt.Errorf("unable to determine categorical maximum bound")
+		}
+		return &numstr.NumberOrString{StrVal: p.Values[len(p.Values)-1], IsString: true}, nil
+	}
+
+	if p.Bounds == nil {
+		return nil, fmt.Errorf("unable to determine numeric maximum bound")
+	}
+
+	return &numstr.NumberOrString{NumVal: p.Bounds.Max}, nil
+}
+
+// ParseValue attempts to parse the supplied value into a NumberOrString based on the type of this parameter.
+func (p *Parameter) ParseValue(s string) (*numstr.NumberOrString, error) {
+	var v numstr.NumberOrString
+	switch p.Type {
+	case ParameterTypeInteger:
+		if _, err := strconv.ParseInt(s, 10, 64); err != nil {
+			return nil, err
+		}
+		v = numstr.FromNumber(json.Number(s))
+	case ParameterTypeDouble:
+		if _, err := strconv.ParseFloat(s, 64); err != nil {
+			return nil, err
+		}
+		v = numstr.FromNumber(json.Number(s))
+	case ParameterTypeCategorical:
+		v = numstr.FromString(s)
+	}
+	return &v, nil
+}
+
+// CheckParameterValue validates that the supplied value can be used for a parameter.
+func CheckParameterValue(p *Parameter, v *numstr.NumberOrString) error {
+	if p.Type == ParameterTypeCategorical {
```

Comment:
Should we add this to the switch statement below?

Reply:
This if statement handles the string case (we can't make it down to the switch, the `IsString` check would fail), the switch below is for the numeric checks. This if block ends with a return.